rfcs/2021-08-06-8619-framing-and-codecs-sources.md

```diff
@@ -0,0 +1,212 @@
+stream of bytes into byte frames (chunks of data that have finite size and
+contain complete messages), while a _Codec_ converts between a byte frame and
+structured data.
+
+Conceptually, we want reusable pieces of logic which would allow us to collapse
+a component + transform into a component with codec config in places where the
+transform was merely used to convert between encoding formats, e.g. from `bytes`
+to `json` or vice versa.
+
+## Scope
+
+For this part, we focus on the source/decoding side and leave the sink/encoding
+side to a separate RFC.
+
+The scope of this RFC concerns at which level framing and decoding operates, how
+these framers and decoders can be configured, and how they can be shared in a
+uniform way. It does not cover any specific implementation for a framer or a
+decoder. Also see [future work](#future-work) for extended goals that are not
+covered by this RFC.
+
+## Motivation
+
+Currently, we have no explicit abstraction that is responsible for handling
+framing and decoding, such that each source may implement these in their own
+way. This has led to inconsistencies in available options and poses additional
+maintenance burden. Components have their own defaults, making
+behavior unpredictable and surprising users, e.g. as documented in
+[#3453](https://github.com/timberio/vector/issues/3453).
+
+## Internal Proposal
+
+To expose framing and decoding to sources in a uniform way, we specify a common
+configuration struct:
+
+```rust
+pub struct DecodingConfig {
+    framing: Option<FramingConfig>,
+    decoding: Option<ParserConfig>,
+}
+```
+
+while the framing method and parser used to deserialize into a structured event
+can be selected from an enum:
+
+```rust
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum FramingConfig {
+    NewlineDelimited,
+    CharacterDelimited {
+        delimiter: char,
+        max_length: Option<usize>,
+    },
+    OctetCounting {
+        max_length: Option<usize>,
+    },
+    ...
+}
+
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum ParserConfig {
+    Bytes,
+    Json {
+        ...
+    },
+    Syslog,
+    ...
+}
+```
+
+The `DecodingConfig` exposes a `build` method to create a Vector `Decoder` which
+implements the `tokio_util::codec::Decoder<Item = (Event, usize)>` trait. That
+way we can produce `Event`s from `ByteMut` by repeated calls to
+`decode`/`decode_eof` either on a byte stream or a byte message. The additional
+`usize` item conveys how many bytes were read to produce the particular `Event`,
+so that this information can be passed along to e.g. the internal event log.
+
+Internally, the `Decoder` holds a framer and a parser:
+
+```rust
+pub struct Decoder {
+    framer: Box<dyn tokio_util::codec::Decoder<Item = Bytes> + Send + Sync>,
+    parser: Box<dyn Parser + Send + Sync>,
+}
+```
+
+with the `Parser` trait being defined as:
+
+```rust
+pub trait Parser {
+    fn parse(&self, bytes: Bytes) -> crate::Result<Event>;
+}
```

Comment:
Also related to remap, we should think about ways to reduce duplication between these parsers and remap `parse_` functions. For example, having one place to define them and direct mapping between function arguments and toml options.

Reply:
Added in ab9c9ebc505582fff656725efe2210df22d380b4